feat: Add support to audit and view audit log with Booking CREATED action

hariombalhara opened the pull request:
## What does this PR do?

This PR introduces the booking audit infrastructure with a viewer service to fetch, enrich, and format audit logs. This is the first iteration that includes only the CREATED action with versioned schema

Note: Additional audit actions, UI improvements to match with Figma will be taken care of in followup PRs, to ensure the PR size remains as small as possible(as it is large enough already)

Demo - Loom - https://www.loom.com/share/22c2f38b052b480b8be3716e1608eaf6

### Key Changes

**New Booking Audit Package** (`packages/features/booking-audit/`):

- `BookingAuditViewerService` - Fetches, enriches, and formats audit logs for display
- `BookingAuditTaskConsumer` - Processes audit tasks asynchronously via Tasker
- `CreatedAuditActionService` (v1) - Handles RECORD_CREATED action with versioned schema

**Repository Layer**:

- `BookingAuditRepository` - CRUD operations for audit records
- `AuditActorRepository` - Actor management (users, guests, system)
- `UserRepository.findByUuid()` - User lookup for actor enrichment

**UI Components**:

- New page at `/booking/logs/[bookinguid]` for viewing audit history
- Filterable audit log list with search, type, and actor filters
- Expandable log entries showing detailed change information

**Infrastructure**:

- `bookingAudit` Tasker task type for async processing
- `booking-audit` feature flag (disabled by default)
- tRPC endpoint `viewer.bookings.getAuditLogs`

### Updates since last revision

- **Refactored `Task` class to `TaskRepository`**: Converted all static methods to instance methods following the repository pattern. A singleton `Task` is exported for backward compatibility, so existing call sites continue to work unchanged.

### Important Notes for Reviewers

- **Permission check is TODO**: `checkPermissions()` in `BookingAuditViewerService` is not yet implemented
- **Only CREATED action supported**: Other actions will throw an error - additional actions will be added iteratively in follow-up PRs
- **Feature flag controlled**: System is behind `booking-audit` flag, disabled by default
- **UI strings need i18n**: Some UI text in `booking-logs-view.tsx` may need translation

### Human Review Checklist

- [ ] Verify permission enforcement strategy for audit log access
- [ ] Review DI wiring in module files
- [ ] Confirm error handling in `BookingAuditTaskConsumer` is appropriate for retry scenarios
- [ ] Check if UI strings need to be added to translation files
- [ ] Verify `TaskRepository` refactoring maintains backward compatibility (singleton export `Task` should work for all existing call sites)

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A - internal infrastructure
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Run the integration tests:

```bash
TZ=UTC yarn test packages/features/booking-audit/lib/service/BookingAuditViewerService.integration-test.ts
TZ=UTC yarn test packages/features/booking-audit/lib/service/BookingAuditTaskConsumer.integration-test.ts
```

To test the UI:

1. Enable the `booking-audit` feature flag in the database
2. Create a booking (only CREATED action is supported in this PR)
3. Navigate to `/booking/logs/{bookingUid}` to view the audit trail

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

<!-- Link to Devin run: https://app.devin.ai/sessions/a8ae61c253b549429401c9651dcbcf44 -->
<!-- Requested by: hariom@cal.com (@hariombalhara) -->

hariombalhara committed:
feat: Add BookingAuditViewerService demo with CREATED action only

- Add CreatedAuditActionService with schema versioning
- Add BookingAuditViewerService supporting only CREATED action for demo
- Add integration tests for write-read cycle
- Add repository read operations (findAllForBooking)
- Add DI infrastructure for viewer service
- Add integration example in BookingAuditService JSDoc
- Add i18n keys documentation

Note: This is a minimal demo. Full action support exists in booking-audit-more-infra branch.

hariombalhara committed:
feat: Implement Booking Audit System with Tasker Integration

- Add BookingAuditTaskConsumer and BookingAuditProducerService for handling booking audit tasks.
- Introduce BookingAuditViewerService for retrieving enriched audit logs.
- Create new schemas and types for booking audit actions and payloads.
- Implement integration tests for the booking audit system.
- Update DI infrastructure to support new services and modules.
- Enhance booking event handling to queue audit tasks on booking creation.

This commit lays the foundation for a comprehensive booking audit system, enabling detailed tracking and retrieval of booking actions.

hariombalhara committed:
refactor: Enhance BookingLogsView with improved filtering and display features

- Introduced FilterSearchField and Select components for better search and filtering options in booking logs.
- Updated state management to handle multiple expanded log entries and JSON data visibility.
- Refactored action icon rendering to use string identifiers for icons instead of JSX elements.
- Improved layout and styling for audit log entries, including a timeline view and enhanced detail display.

These changes improve the user experience by providing more intuitive filtering and a clearer presentation of booking audit logs.

hariombalhara committed:
feat: Introduce TaskerService for improved task management

- Added TaskerService to encapsulate tasker functionality and provide a consistent interface for retrieving taskers.
- Updated BookingAuditProducerService to utilize TaskerService for tasker retrieval, enhancing dependency injection.
- Refactored imports in regular-booking.module.ts and bookings.ts to include the new TaskerService.

These changes streamline task management and improve the modularity of the booking audit system.

Udit-takkar commented on packages/features/booking-audit/lib/service/booking-audit.integration-test.ts:
```diff
@@ -0,0 +1,284 @@
+
+      // Verify audit data matches (getDisplayJson returns fields only, not versioned wrapper)
+      const displayData = auditLog.data as { startTime: number; endTime: number; status: string };
+      expect(displayData.startTime).toBe(booking!.startTime.getTime());
```
Integration tests are failing

hariombalhara replied:
Fixed it. last moment change broke it. Thanks !!

hariombalhara commented on apps/web/modules/booking/logs/views/booking-logs-view.tsx:
```diff
@@ -0,0 +1,303 @@
+    }
+};
+
+export default function BookingLogsView({ bookingUid }: BookingLogsViewProps) {
```
https://www.figma.com/design/wleA2SR6rn60EK7ORxAfMy/Cal.com-New-Features?node-id=6878-92169&t=Pk1ex8fGI01srpra-0

This is the figma, I have already converted the timestamp. Will followup with improvements in a separate PR.

Regarding Actor's email and phone, I don't think designs have it.

hariombalhara commented on packages/features/booking-audit/lib/service/BookingAuditViewerService.ts:
```diff
@@ -0,0 +1,176 @@
+        _userEmail: string
+    ): Promise<{ bookingUid: string; auditLogs: EnrichedAuditLog[] }> {
+        // Check permissions
+        await this.checkPermissions();
```
Yeah, I need to fully use PBAC here, I will followup for that in a separate PR. Right now in production mode checkPermisions simply don't allow accessing audit logs.

hariombalhara committed:
refactor: Enhance BookingLogsView with new filtering components and improved structure

- Introduced BookingLogsFilters and BookingLogsTimeline components to separate concerns and improve readability.
- Updated state management for search and filter functionality, allowing for better user interaction.
- Refactored the rendering of audit logs to enhance layout and detail presentation.
- Added type definitions for audit logs and filters to improve type safety.

These changes streamline the booking logs view, making it more modular and user-friendly.

volnei commented on packages/features/tasker/repository.ts:
```diff
@@ -142,6 +142,23 @@ export class Task {
+      data: {
+        payload: newPayload,
+      },
+    });
```
@hariombalhara we're avoiding static repository methods, can you refactor it please?

volnei reviewed (dismissed):
Let's avoid static methods for repositories 🙏🏼

joeauyeung commented on packages/features/booking-audit/lib/repository/PrismaAuditActorRepository.ts:
```diff
@@ -25,5 +25,87 @@ export class PrismaAuditActorRepository implements IAuditActorRepository {
+        }
+
+        // First try to find by email if email exists
+        if (normalizedEmail) {
```
We should handle the scenario if an actor has both an email and phone values. Right now it'll make 4 possible DB calls because of the two `if` statements

hariombalhara replied:
Let me think about how to handle it well.

joeauyeung commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -412,6 +414,49 @@ function formatAvailabilitySnapshot(data: {
   };
 }
 
+function buildBookingCreatedPayload({
```
I'm hesitant about having this here in `RegularBookingService` it feels like it should live as a method in another service.

hariombalhara replied:
So, this is building all the data that is needed to fire an event(bookingCreated). This is usually the responsibility of the calling site itself because it knows where to get the data from and then only thing left is to properly place it as required by event handler signature. Type safety ensures they are passed correctly.

joeauyeung commented on packages/features/bookings/lib/service/RegularBookingService.ts:
```diff
@@ -2243,32 +2288,36 @@ async function handler(
   } else {
-    await bookingEventHandler.onBookingCreated(bookingCreatedPayload);
+    // TODO: We need to check session in booking flow and accordingly create USER actor if applicable.
+    const auditActor = makeGuestActor({ email: bookerEmail, name: fullName });
```
Likewise should this be handled in another service? Maybe the one on the line below? I want to avoid needing people to call these specific functions in different features.

hariombalhara replied:
I am not so sure about this as well. If you look at this definition, it isn't doing much at all.
My reasoning was that everywhere one would need to define these actors,  and it would be a lot of places. Either we pass the as a POJO or we have a well defined utility fn to create it, so that it is difficult to make mistakes

<img width="581" height="328" alt="Screenshot 2025-12-06 at 8 31 34 AM" src="https://github.com/user-attachments/assets/d253b191-cd4e-4b65-bc9b-bb6f83e9b3cf" />

joeauyeung reviewed (dismissed):
Love the architecture here. This `consumer` and `producer` pattern, is this something we want to adopt throughout the whole codebase?

All my comments are non-blocking but can be addressed in a follow up.

devin-ai-integration[bot] committed:
refactor: Convert Task class to TaskRepository with instance methods

- Renamed Task class to TaskRepository
- Converted all static methods to instance methods
- Added Dependencies type with prismaClient injection
- Exported singleton 'Task' for backward compatibility
- All existing call sites continue to work unchanged

Co-Authored-By: hariom@cal.com <hariombalhara@gmail.com>

hariombalhara committed:
fix failing tests

hariombalhara commented on apps/web/app/(use-page-wrapper)/(main-nav)/booking/[uid]/logs/page.tsx:
Yeah, I have been thinking about it. So right now I haven;t implemented permission check at all.

I believe audit logs is initially for admins only and they need to see all the data.

Going forward we might need to show audit logs for regular members too, and there we might need to hide certain details.

hariombalhara committed:
Merge branch 'main' into booking-audit-viewer-demo

ThyMinimalDev reviewed (approved):
API v2 side looks good
